modules/service/src/main/scala/lucuma/odb/service/ItcService.scala

```diff
@@ -0,0 +1,379 @@
+      programId:     Program.Id,
+      observationId: Observation.Id
+    ) extends Error {
+      def format: String =
+        s"Observation '$observationId' in program '$programId' not found."
+    }
+
+    case class MissingParams(
+      missing: NonEmptyList[GeneratorParamsService.MissingData]
+    ) extends Error {
+      def format: String = {
+        val params = missing.map { m =>
+          s"${m.targetId.fold("") { tid => s"(target $tid) " }}${m.paramName}"
+        }.intercalate(", ")
+        s"ITC cannot be queried until the following parameters are defined: $params"
+      }
+    }
+
+    case class RemoteServiceErrors(
+      problems: NonEmptyList[(Target.Id, String)]
+    ) extends Error {
+      def format: String = {
+        val ps = problems.map { case (tid, msg) => s"Target '$tid': $msg"}
+        s"ITC returned errors: ${ps.intercalate(", ")}"
+      }
+    }
+  }
+
+  case class TargetResult(
+    targetId: Target.Id,
+    input:    SpectroscopyIntegrationTimeInput,
+    value:    IntegrationTime
+  ) {
+    def totalTime: Option[TimeSpan] = {
+      val total = BigInt(value.exposureTime.toMicroseconds) * value.exposures.value
+      Option.when(total.isValidLong)(TimeSpan.fromMicroseconds(total.longValue)).flatten
+    }
+  }
+
+  object TargetResult {
+
+    given Order[TargetResult] =
+      Order.by { s => (s.totalTime, s.targetId) }
+
+    import lucuma.odb.json.time.query.given
+
+    given Encoder[TargetResult] =
+      Encoder.instance { s =>
+        Json.obj(
+          "targetId"      -> s.targetId.asJson,
+          "exposureTime"  -> s.value.exposureTime.asJson,
+          "exposures"     -> s.value.exposures.value.asJson,
+          "signalToNoise" -> s.value.signalToNoise.asJson
+        )
+      }
+  }
+
+  case class AsterismResult(
+    value: Zipper[TargetResult]
+  )
+
+  object AsterismResult {
+
+    given Encoder[AsterismResult] =
+      Encoder.instance { rs =>
+        Json.obj(
+          "result" -> rs.value.focus.asJson,
+          "all"    -> rs.value.toList.asJson
+        )
+      }
+
+  }
+
+  case class Success(
+    params: GeneratorParams,
+    result: AsterismResult
+  )
+
+  private val hex = HexFormat.of
+
+  private def hash(input: SpectroscopyIntegrationTimeInput): String =
+    hex.formatHex(
+      MessageDigest
+        .getInstance("MD5")
+        .digest(input.asJson.noSpaces.getBytes("UTF-8"))
+    )
+
+  def pollVersionsForever[F[_]: Async: Temporal](
+    client:     Resource[F, ItcClient[F]],
+    session:    Resource[F, Session[F]],
+    pollPeriod: FiniteDuration
+  ): F[Unit] = {
+    val pollOnce: F[Unit] =
+      for {
+        v <- client.use(_.versions)
+        _ <- session.use { s => s.transaction.use(_ => s.execute(Statements.UpdateItcVersion)(v.server.some, v.data)) }
+      } yield ()
+
+    Stream
+      .fixedDelay(pollPeriod)
+      .zip(Stream.repeatEval(pollOnce))
+      .compile
+      .drain
+      .start
+      .void
```

Comment:
* Is `.start` the right way to do this?  I want it to continue as long as the process is running.
* Note `pollVersionsForever` is called on startup and there's no associated user or services.  A session created via `Session.single` is used for this.
* I suppose this will be done in all instances of the database server.  Is this an issue?

Reply:
I  think it is fine to have it in all instances as they need to clean their local caches

Reply:
I think `start` is fine here. If you needed a finalizer on shutdown you could use `Supervisor` but I don't think you need it.